internal/util-logging/src/main/scala/sbt/internal/util/Terminal.scala

```diff
@@ -0,0 +1,210 @@
+import java.util.Locale
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicReference }
+import java.util.concurrent.locks.ReentrantLock
+
+import jline.console.ConsoleReader
+
+import scala.util.control.NonFatal
+
+object Terminal {
+
+  /**
+   * Gets the current width of the terminal. The implementation reads a property from the jline
+   * config which is updated if it has been more than a second since the last update. It is thus
+   * possible for this value to be stale.
+   *
+   * @return the terminal width.
+   */
+  def getWidth: Int = terminal.getWidth
+
+  /**
+   * Gets the current height of the terminal. The implementation reads a property from the jline
+   * config which is updated if it has been more than a second since the last update. It is thus
+   * possible for this value to be stale.
+   *
+   * @return the terminal height.
+   */
+  def getHeight: Int = terminal.getHeight
+
+  /**
+   * Returns true if the current terminal supports ansi characters.
+   *
+   * @return true if the current terminal supports ansi escape codes.
+   */
+  def isAnsiSupported: Boolean =
+    try terminal.isAnsiSupported
+    catch { case NonFatal(_) => !isWindows }
+
+  /**
+   * Returns true if System.in is attached. When sbt is run as a subprocess, like in scripted or
+   * as a server, System.in will not be attached and this method will return false. Otherwise
+   * it will return true.
+   *
+   * @return true if System.in is attached.
+   */
+  def systemInIsAttached: Boolean = attached.get
+
+  /**
+   * Provides a wrapper around System.in. The wrapped stream in will check if the terminal is attached
+   * in available and read. If a read returns -1, it will mark System.in as unattached so that
+   * it can be detected by [[systemInIsAttached]].
+   *
+   * @return the wrapped InputStream
+   */
+  private[sbt] def wrappedSystemIn: InputStream = WrappedSystemIn
+
+  /**
+   * Restore the terminal to its initial state.
+   */
+  private[sbt] def restore(): Unit = terminal.restore()
+
+  /**
+   * Runs a thunk ensuring that the terminal has echo enabled. Most of the time sbt should have
+   * echo mode on except when it is explicitly set to raw mode via [[withRawSystemIn]].
+   *
+   * @param f the thunk to run
+   * @tparam T the result type of the thunk
+   * @return the result of the thunk
+   */
+  private[sbt] def withEcho[T](toggle: Boolean)(f: => T): T = {
+    val previous = terminal.isEchoEnabled
+    terminalLock.lockInterruptibly()
+    try {
+      terminal.setEchoEnabled(toggle)
+      f
+    } finally {
+      terminal.setEchoEnabled(previous)
+      terminalLock.unlock()
+    }
+  }
+
+  /**
+   * Runs a thunk ensuring that the terminal is in canonical mode:
+   * [[https://www.gnu.org/software/libc/manual/html_node/Canonical-or-Not.html Canonical or Not]].
+   * Most of the time sbt should be in canonical mode except when it is explicitly set to raw mode
+   * via [[withRawSystemIn]].
```

Comment:
```suggest
   * sbt should go back to canonical mode before giving the controls to `run` or `sbt new`.
```

Reply:
Not sure I agree with this suggestion. sbt should only ever _not_ be in canonical mode when in a `Terminal.withRawSystemIn` block so the onus should not actually be on the caller to go back to canonical mode.